### DISCUSSION TITLE
How can I implement Text input components?

### DISCUSSION THREAD

**Discussion by kyuranger** [AvaloniaUI/Avalonia#15152] [Q&A] [answered]
Recently I am working hard on virtual keyboard in Avalonia Linux DRM. 

Although I do it via some simple events like:
```
InputElement.GotFocusEvent.AddClassHandler<TextBox>(GotFocusTextBox);
InputElement.LostFocusEvent.AddClassHandler<TextBox>(LostFocusTextBox);
```

![1](https://github.com/AvaloniaUI/Avalonia/assets/30720036/0bab4c77-fc9d-4735-be5f-32adb098666b)

But I don’t implement it through Avalonia’s official interface(https://github.com/AvaloniaUI/Avalonia/wiki/IME).

I hope my input method can still follow the official interface of Avalonia.

However, I don’t understand how to use Avalonia’s official interface at all, and there are no examples or tutorials.

![image](https://github.com/AvaloniaUI/Avalonia/assets/30720036/0eeed014-8761-497f-8eb5-febf74a4eac3)

And I checked some related source codes and found that some interfaces were marked as Unstable, and I was even more unclear about how to use them.

Please tell me how to implement it according to the Avalonia official interface (it is best to have relevant examples or tutorials).

**Comment by maxkatz6** [comment, marked as answer]
You are looking at wrong interfaces. These two are internal implementation details, that are provided by the platform backends (ios, android...). You can't implement these.

If you develop a custom TextBox-like control, you need to use TextBoxTextInputMethodClient:
https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Controls/TextBox.cs#L342-L348
https://github.com/AvaloniaUI/AvaloniaEdit/blob/adbe72b8023207270d30c16beb4fadcd6d598023/src/AvaloniaEdit/Editing/TextArea.cs#L71-L77

But since you develop a custom keyboard, while keeping default input controls, you likely don't need that. Just send events to the focused control.

**Reply by kyuranger** [reply]
Thank you so much. By the way, someone said that I should access the TopLevel and send events to the focused control.

However, my way is to set the InputElement.GotFocusEvent/InputElement.LostFocusEvent that different from it.

Although my method feels feasible, I'm not sure if there are any hidden dangers in doing so. Could you please tell me the answer?

**Reply by maxkatz6** [reply]
Well, I don't have experience building a virtual keyboard. But the elephant in the room is support of all kinds of different languages and input methods in the keyboard itself. If you only need English, it shouldn't be much difficult.

**Reply by kyuranger** [reply]
Thanks a lot.
